src/ast_util.py

```diff
@@ -166,6 +166,126 @@ def generic_visit(self, node: ca.Node) -> None:
+    for item in ast.ext:
+        if isinstance(item, ca.Decl) and item.name:
+            used_names.add(item.name)
+        if isinstance(item, ca.FuncDef) and item.decl.name:
+            used_names.add(item.decl.name)
+    return get_noncolliding_name(used_names, name)
+
+
+def get_noncolliding_struct_name(typemap: TypeMap, name: str) -> str:
+    used_names: Set[str] = set()
+    for struct_name in typemap.struct_defs.keys():
+        used_names.add(struct_name)
+    for type_name in typemap.typedefs.keys():
+        used_names.add(type_name)
+    return get_noncolliding_name(used_names, name)
+
+
+def get_noncolliding_name(used_names: Set[str], name: str) -> str:
+    new_name = name
+    counter = 1
+    while new_name in used_names:
+        counter += 1
+        new_name = f"{new_name}{counter}"
+    return new_name
+
+
+def ast_ext_struct_idx(ast: ca.FileAST, struct: ca.Struct) -> int:
+    """Returns the index of the node in `ast.ext` that contains the definition
+    of `struct`.
+    """
+    found = False
+    class StructVisitor(ca.NodeVisitor):
+        def visit_Struct(self, node: ca.Struct) -> None:
+            nonlocal found
+            if node is struct:
+                found = True
+            self.generic_visit(node)
+
+    for i, node in enumerate(ast.ext):
+        StructVisitor().visit(node)
+        if found:
+            return i
+    return -1
+
+
+def copy_struct(ast: ca.FileAST, typemap: TypeMap, struct: ca.Struct) -> ca.Struct:
+    """Adds a new struct with a definition matching `struct` to the AST.
+    Returns this new struct.
+    """
+    new_struct = copy.deepcopy(struct)
+    base_name = struct.name if struct.name else struct.decls[0].name
+    new_struct.name = get_noncolliding_struct_name(typemap, base_name)
+
+    # Insert the new struct right above the base struct. This preserves any
+    # struct field dependencies.
+    insert_idx = ast_ext_struct_idx(ast, struct)
+    decl = ca.Decl(name=None, quals=[], align=[], storage=[], funcspec=[], type=new_struct, init=None, bitsize=None)
+    ast.ext = ast.ext[:insert_idx] + [decl] + ast.ext[insert_idx:]
+    return new_struct
+
+
+def struct_set_field_type(parent: ca.Struct, field_idx: int, new_type: ca.Struct) -> None:
+    struct_type = copy.deepcopy(new_type)
+    struct_type.decls = None
+    old_type_decl = parent.decls[field_idx].type
+    assert isinstance(old_type_decl, ca.TypeDecl), "can only modify the type of a TypeDecl field"
+    old_type_decl.type = struct_type
+
+
+def struct_substruct_idx(parent: ca.Struct, substruct: ca.Struct, typemap: TypeMap) -> int:
+    """Returns the index of the field in `parent` of type `substruct`.
+    """
+    for i, decl in enumerate(parent.decls):
+        if isinstance(decl.type, ca.TypeDecl):
+            decl_struct = resolve_struct_def(decl.type, typemap)
+            if decl_struct is substruct:
+                return i
+    return -1
+
+
+def struct_rename_field(field: ca.Decl, name: str) -> None:
```

Comment:
This is the same as `set_decl_name` more or less

Reply:
Ah thanks, I switched to `set_decl_name()` here